File: javapatterns/src/main/java/org/iut/nantes/factories/pizzas/abstractfactory/Pizza.java
package org.iut.nantes.factories.pizzas.abstractfactory;

abstract class Pizza {
    protected CreerIngredients creerIngredients;

    protected Pate pate;
    protected Sauce sauce;
    protected Fromage fromage;

    public Pizza(CreerIngredients creerIngredients) {
        this.creerIngredients = creerIngredients;
    }

    abstract void preparer();
}

class PizzaRoyale extends Pizza {

    public PizzaRoyale(CreerIngredients creerIngredients) {
        super(creerIngredients);
    }

    @Override
    void preparer() {
        pate = creerIngredients.creerPate();
        sauce = creerIngredients.creerSauce();
        fromage = creerIngredients.creerFromage();
    }

}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/factories/pizzas/abstractfactory/Pizza.kt
package org.iut.nantes.factories.pizzas.abstractfactory

abstract class Pizza(protected var creerIngredients: CreerIngredients) {
    protected var pate: Pate? = null
    protected var sauce: Sauce? = null
    protected var fromage: Fromage? = null
    abstract fun preparer()
}

class PizzaRoyale(creerIngredients: CreerIngredients) : Pizza(creerIngredients) {
    override fun preparer() {
        pate = creerIngredients.creerPate()
        sauce = creerIngredients.creerSauce()
        fromage = creerIngredients.creerFromage()
    }
}